java/src/main/java/org/signal/libsignal/metadata/protocol/UnidentifiedSenderMessageContent.java

```diff
@@ -32,37 +34,44 @@ public UnidentifiedSenderMessageContent(byte[] serialized) throws InvalidMetadat
         default:             throw new InvalidMetadataMessageException("Unknown type: " + message.getType().getNumber());
       }
 
-      this.senderCertificate = new SenderCertificate(message.getSenderCertificate().toByteArray());
-      this.content           = message.getContent().toByteArray();
-      this.serialized        = serialized;
+      this.senderIdentityKey = message.getSenderIdentity().toByteArray();
+      this.senderDeviceId = message.getSenderDeviceId().toByteArray();
+      this.content        = message.getContent().toByteArray();
+      this.serialized     = serialized;
     } catch (InvalidProtocolBufferException e) {
       throw new InvalidMetadataMessageException(e);
     }
   }
 
-  public UnidentifiedSenderMessageContent(int type, SenderCertificate senderCertificate, byte[] content) {
-    try {
-      this.serialized = SignalProtos.UnidentifiedSenderMessage.Message.newBuilder()
-                                                                      .setType(SignalProtos.UnidentifiedSenderMessage.Message.Type.valueOf(getProtoType(type)))
-                                                                      .setSenderCertificate(SignalProtos.SenderCertificate.parseFrom(senderCertificate.getSerialized()))
-                                                                      .setContent(ByteString.copyFrom(content))
-                                                                      .build()
-                                                                      .toByteArray();
-
-      this.type = type;
-      this.senderCertificate = senderCertificate;
-      this.content = content;
-    } catch (InvalidProtocolBufferException e) {
-      throw new AssertionError(e);
-    }
+  public UnidentifiedSenderMessageContent(int type, byte[] senderIdentityKey, byte[] senderDeviceId, byte[] content) {
+    this.serialized = SignalProtos.UnidentifiedSenderMessage.Message.newBuilder()
+                                                                    .setType(SignalProtos.UnidentifiedSenderMessage.Message.Type.valueOf(getProtoType(type)))
+                                                                    .setSenderIdentity(ByteString.copyFrom(senderIdentityKey))
```

Comment:
Somewhere we still need to verify that the sender identity lines up with the peer's public key.  Did I miss where this happens?

Reply:
This'll make more sense after you see my changes to libsignal-protocol-java, but wrapped signal message no longer contain the sender identity since that's implicit in the address. What ensures that we can't spoof is that the session key of the message is tracked in the local SessionStore and ties to the real sender, not the spoofed one.

Reply:
Going to do some more testing here though :)

Reply:
Okay, I've added some more test cases and added a comment that hopefully makes it clearer why we're okay here.

Reply:
Hmm so are you saying that the sender identity is set somewhere else then?  Why set it here too?  Or it's overwritten later?  To be more succinct - if we can't trust _this_ sender identity, but there's another one which we _can_ trust, then why do we use this one at all?

Or am I misunderstanding?  Is `senderIdentityKey` controllable by the sender?

Reply:
Signal message exchanges are stateful. Each side maintains session state, keyed to the other party's address (id plus device ID). The very first message in an exchange is a session initiation message that includes some additional stuff to allow for a DH key exchange. Once the sender gets back a reply, then the session is fully established. As the clients send messages to each other, they evolve the session keys using a double-ratchet protocol in order to achieve perfect forward secrecy (i.e. even if someone recovers the original session key, they won't be able to read future messages).

So, in order to decrypt a message, a recipient needs to know whom its from so that it can look up the relevant session information in its local session store.

A normal (not sealed sender) Signal message doesn't actually include any information about the sender. So for example, if you have a direct P2P connection with the other party (like with WebRTC), you don't need an envelope that tells you whom you're speaking with, because you already know that.

The envelope is useful when we're ferrying messages through a message broker (tassis), the receiving client needs to know the sender so that they can.

So, the reason that the UnidentifiedSenderMessage (the envelope) contains the sender is so that the receiver knows how to decrypt. However, if the sender doesn't match who actually encrypted the message, decryption will fail because the MAC verification fails.

Reply:
Ah right, the sender identity is needed to look up the decryption keys.  That makes sense.  So this is an unverified sender identity.  Maybe it should be named as such then?  Or we could specify that in a comment?  What do you think @oxtoacart?

Reply:
Because we don't use phone number, we don't verify anything, so in our system you'll never see a "verified" identity.

Reply:
Right, sorry, I don't mean "verified by Lantern".  I mean "securely verified".  This is the sender's public key, right?  But it hasn't been verified at this point in the code if I'm understanding correctly.  Therefore I was suggesting that this be made clear.  My logic is that someone may come by later to maintain or update this code and use this unverified identity inappropriately.